traits/trait_set_object.py

```diff
@@ -263,35 +460,264 @@ def __reduce_ex__(self, protocol=None):
+    Traits versions before 6.0.
+
+    Parameters
+    ----------
+    trait : CTrait
+        The trait that the list has been assigned to.
+    object : HasTraits
+        The object the list belongs to.
+    name : str
+        The name of the trait on the object.
+    value : iterable
+        The initial value of the list.
+    notifiers : list
+        Additional notifiers for the list.
+
+    Attributes
+    ----------
+    trait : CTrait
+        The trait that the list has been assigned to.
+    object : HasTraits
+        The object the list belongs to.
+    name : str
+        The name of the trait on the object.
+    value : iterable
+        The initial value of the list.
+    notifiers : list
+        Additional notifiers for the list.
+    """
+
+    def __init__(self, trait, object, name, value, notifiers=[]):
+
+        self.trait = trait
+        self.object = ref(object)
+        self.name = name
+        self.name_items = None
+        if trait.has_items:
+            self.name_items = name + "_items"
+
+        super().__init__(value, validator=self.validator,
+                         notifiers=[self.notifier] + notifiers)
+
+    def validator(self, current_set, value):
+        """ Validates the value by calling the inner trait's validate method
+        and also ensures that the size of the list is within the specified
+        bounds.
+
+        Parameters
+        ----------
+        current_set : set
+            The current contents of the set.
+        value : set
+            set of values that need to be validated.
+
+        Returns
+        -------
+        value : set of validated values
+
+        Raises
+        ------
+        TraitError
+            On validation failure for the inner trait or if the size of the
+            list exceeds the specified bounds.
+
+        """
+        object_ref = getattr(self, 'object', None)
+        trait = getattr(self, 'trait', None)
+
+        if object_ref is None or trait is None:
+            return value
+
+        object = object_ref()
+
+        # validate the new value(s)
+        validate = trait.item_trait.handler.validate
+        if validate is None:
+            return value
+
+        validate = adapt_trait_validator(validate, name=self.name)
+        return validate(object, value)
+
+    def notifier(self, removed, added):
+        """ Converts and consolidates the parameters to a TraitSetEvent and
+        then fires the event.
+
+        Parameters
+        ----------
+        removed : set
+            Set of values that were removed.
+        added : set
+            Set of values that were added.
+
+        Returns
+        -------
+        None
+
+        """
+        is_trait_none = getattr(self, 'trait', None) is None
+        is_name_items_none = getattr(self, 'name_items', None) is None
+        is_object_ref_none = getattr(self, 'object', None) is None
+        if is_trait_none or is_name_items_none or is_object_ref_none:
+            return
```

Comment:
Had to add this to `TraitList` to get some tests to pass in python 3.8. Need to recreate the equivalent tests here and check.

Reply:
It looks like the pickle-unpickle test which had failed for `TraitList` in Python 3.8 also includes a test for `Set`. 
Ref: traits.tests.test_trait_list_dict.TestTraitListDictSetPersistence.test_pickle_whole

I've verified that the test does not fail for python 3.8, removing this block.

Reply:
I tried removing this block of code and no tests fail for Python 3.8

If no tests fail when this block of code is removed, it indicates one of these two:
(1) This block of code is redundant and should be removed.
(2) This block of code is needed but we are lacking the relevant test, and we should add a test

Both of these options require actions. I will investigate option 2.

Reply:
On `self.object` and `self.trait`, looks like the minimum we need is just this:
```
        object = self.object()
        if object is None:
            return
```

This is for two use cases:
1) When the set is pickled and unpickled on its own. In that case, both `trait` and `object` are set to None. A set unpickled on its own cannot fire notifications. Checking either `object is None` or `self.trait is None` is enough.
2) When the object is deep-copied, the set's object reference is lost, but the trait is still there.

I can't think of a scenario where `object` is not None but `trait` is None.

On `self.name_items`, that is because the `List`/`Set`/`Dict` trait has a flag for switching off items notifications altogether:
```
Set(items=False)
```
Honestly I don't know why such a feature exists, but it is there and the check on `if self.name_items is not None` is all over the place.
We are missing the test for `name_items` being None.
Also, I don't think we need the lenient `getattr(self, "name_items", None)`, getting the attribute directly with `self.name_items` should be fine, right?

Reply:
> We are missing the test for name_items being None.

I am on it... PR coming...

Reply:
#990

Reply:
Resolved via #990